addons/marketing_campaign/models/marketing_campaign.py

```diff
@@ -0,0 +1,122 @@
+Test in Realtime - It creates and processes all the activities directly but does not send emails or produce reports.
+With Manual Confirmation - the campaigns runs normally, but the user has to validate all workitem manually.
+Normal - the campaign runs normally and automatically sends all emails and reports (be very careful with this mode, you're live!)""")
+    state = fields.Selection([('draft', 'New'),
+                            ('running', 'Running'),
+                            ('cancelled', 'Cancelled'),
+                            ('done', 'Done')], string='Status', copy=False, default='draft')
+    activity_ids = fields.One2many('marketing.campaign.activity', inverse_name='campaign_id', string='Activities')
+    fixed_cost = fields.Float(help="Fixed cost for running this campaign. You may also specify variable cost and revenue on each campaign activity. Cost and Revenue statistics are included in Campaign Reporting.", digits=dp.get_precision('Product Price'))
+    segment_ids = fields.One2many('marketing.campaign.segment', inverse_name='campaign_id', string='Segments', readonly=False)
+    segments_count = fields.Integer(compute='_compute_segments_count', string='Segments')
+
+    def _compute_segments_count(self):
+        segments = self.env['marketing.campaign.segment'].read_group([('campaign_id', 'in', self.ids)], ['campaign_id'], ['campaign_id'])
+        mapped_data = dict([(m['campaign_id'][0], m['campaign_id_count']) for m in segments])
+        for campaign in self:
+            campaign.segments_count = mapped_data.get(campaign.id, 0)
+
+    def _get_partner_for(self, record):
+        partner_field = self.partner_field_id.name
+        if partner_field:
+            return getattr(record, partner_field)
+        elif self.object_id.model == 'res.partner':
+            return record
+
+    @api.multi
+    def state_running_set(self):
+        # TODO check that all subcampaigns are running
+        self.ensure_one()
+
+        if not self.activity_ids:
+            raise UserError(_("The campaign cannot be started. There are no activities in it."))
+
+        has_start = False
+        has_signal_without_from = False
+
+        for activity in self.activity_ids:
+            if activity.start:
+                has_start = True
+            if activity.signal and len(activity.from_ids) == 0:
+                has_signal_without_from = True
+
+        if not has_start and not has_signal_without_from:
+            raise UserError(_("The campaign cannot be started. It does not have any starting activity. Modify campaign's activities to mark one as the starting point."))
+
+        return self.write({'state': 'running'})
+
+    @api.multi
+    def state_done_set(self):
+        # TODO check that this campaign is not a subcampaign in running mode.
+        segments = self.env['marketing.campaign.segment'].search(
+            [('campaign_id', 'in', self.ids), ('state', '=', 'running')])
```

Comment:
What if I consider self.segment_ids and use filtered for running segments instead of search ?